core/src/main/java/org/elasticsearch/index/query/support/InnerHitBuilder.java

```diff
@@ -0,0 +1,603 @@
+        return this;
+    }
+
+    public InnerHitBuilder addScriptField(String name, Script script) {
+        if (scriptFields == null) {
+            scriptFields = new ArrayList<>();
+        }
+        scriptFields.add(new ScriptField(name, script, false));
+        return this;
+    }
+
+    public FetchSourceContext getFetchSourceContext() {
+        return fetchSourceContext;
+    }
+
+    public InnerHitBuilder setFetchSourceContext(FetchSourceContext fetchSourceContext) {
+        this.fetchSourceContext = fetchSourceContext;
+        return this;
+    }
+
+    public List<SortBuilder<?>> getSorts() {
+        return sorts;
+    }
+
+    public InnerHitBuilder setSorts(List<SortBuilder<?>> sorts) {
+        this.sorts = sorts;
+        return this;
+    }
+
+    public InnerHitBuilder addSort(SortBuilder sort) {
+        if (sorts == null) {
+            sorts = new ArrayList<>();
+        }
+        sorts.add(sort);
+        return this;
+    }
+
+    public HighlightBuilder getHighlightBuilder() {
+        return highlightBuilder;
+    }
+
+    public InnerHitBuilder setHighlightBuilder(HighlightBuilder highlightBuilder) {
+        this.highlightBuilder = highlightBuilder;
+        return this;
+    }
+
+    public QueryBuilder<?> getQuery() {
+        return query;
+    }
+
+    public InnerHitBuilder setQuery(QueryBuilder<?> query) {
+        this.query = query;
+        return this;
+    }
+
+    public InnerHitBuilder setInnerHitsBuilder(InnerHitsBuilder innerHitsBuilder) {
+        this.innerHitsBuilder = innerHitsBuilder;
+        return this;
+    }
+
+    public InnerHitsContext.BaseInnerHits buildInline(SearchContext parentSearchContext, QueryShardContext context) throws IOException {
+        InnerHitsContext.BaseInnerHits innerHitsContext;
+        if (nestedPath != null) {
+            ObjectMapper nestedObjectMapper = context.getObjectMapper(nestedPath);
+            ObjectMapper parentObjectMapper = context.nestedScope().getObjectMapper();
+            innerHitsContext = new InnerHitsContext.NestedInnerHits(
+                    parentSearchContext, parentObjectMapper, nestedObjectMapper
+            );
+        } else if (parentChildType != null) {
+            DocumentMapper documentMapper = context.getMapperService().documentMapper(parentChildType);
+            innerHitsContext = new InnerHitsContext.ParentChildInnerHits(
+                    parentSearchContext, context.getMapperService(), documentMapper
+            );
+        } else {
+            throw new IllegalStateException("Neither a nested or parent/child inner hit");
+        }
+        setupInnerHitsContext(context, innerHitsContext);
+        return innerHitsContext;
+    }
+
+    public InnerHitsContext.BaseInnerHits buildTopLevel(SearchContext parentSearchContext, QueryShardContext context,
+                                                        InnerHitsContext innerHitsContext) throws IOException {
+        if (nestedPath != null) {
```

Comment:
I tried to make this change, but it isn't really possible to have two constructors having both a single string parameter. Also because the object parser is used to parse the xcontent into a builder, no constructor arguments can be passed. The object parser can only work with setters.